### PR TITLE
Make Z3 an optional dependency

### PR DESCRIPTION
**Pull request by AltGr**
If Catala is compiled without Z3, trying to run it with the backend `Proof`
will yield:
```
[ERROR] This instance of Catala was compiled without Z3 support.
```
and return 124

Note that this doesn't change the `make depends`, opam file or CI to account
for it, it just enables it at the build-system level.

There are also no hooks at this moment to have Catala self-document the
options whith which it was compiled (e.g. in the `--help` screen). But that
could be added in a more general way later, it's probably not really needed
yet.

### CODE REVIEW

**Review by denismerigoux — CHANGES_REQUESTED** [review]
Amazing! Altough I would prefer if you didn't rename the `z3backend.ml` into the `z3backend.real.ml`, the "realness" is implicit enough I guess.

Also could you provide some documentation about what `dune` command to run? Maybe create a new target in the top-level `Makefile` ?

**Comment by AltGr** [comment]
> Altough I would prefer if you didn't rename the z3backend.ml into the z3backend.real.ml, the "realness" is implicit enough I guess.

It's exactly what I tried to do at first… but dune enforces it unfortunately :shrug: 

> Also could you provide some documentation about what dune command to run? Maybe create a new target in the top-level Makefile ?

well it's not changed, just running `make build` will work as before, but if `z3` wasn't installed you will still be able to compile. Or did you mean something like a `make depends-except-z3` target ?

**Comment by denismerigoux** [comment]
Yeah an explicit `make depends-except-z3` would be perfect.

**Comment by denismerigoux** [comment]
BTW this should also fix https://github.com/CatalaLang/catala/issues/207, you might want to use the `make depends-except-z3` to build the JS executable of the compiler.

**Comment by AltGr** [comment]
hm wait the CI still installs z3

File: compiler/verification/dune [hunk 1]
```diff
@@ -1,7 +1,17 @@
...
+ (libraries
+  bindlib
+  utils
+  dcalc
+  runtime
+  calendar
+  (select
+   z3backend.ml
+   from
+   (z3 -> z3backend.real.ml)
```
Comment (denismerigoux) [inline compiler/verification/dune:13]: Can you here make the case disjunction based on the fact that the user has passed a particular compilation flag or profile?

**Comment by denismerigoux** [comment]
So there are two types of CI: `run-make-all.yml` is executed on stateful CI managed by Inria on which `z3` has been preinstalled. `run-build.yml` in run on GitHub runners that build everything from scratch at each CI run.

**Review by denismerigoux — APPROVED** [review]

**Pull request merged**